common_patches/usb0.yaml

```diff
@@ -0,0 +1,395 @@
+          name: SUPCNT
+    DOEPDMA0:
+      name: DOEPDMA
+      _modify:
+        DMAADDR0:
+          name: DMAADDR
+    DOEPDMAB0:
+      name: DOEPDMAB
+      _modify:
+        DMABUFFERADDR0:
+          name: DMABUFFERADDR
+  OUT_EP%s:
+    description: "Device OUT endpoints 1-6"
+    DOEPCTL[1-6]:
+      name: DOEPCTL
+    DOEPINT[1-6]:
+      name: DOEPINT
+      _modify:
+        XFERCOMPL1:
+          name: XFERCOMPL
+        EPDISBLD1:
+          name: EPDISBLD
+        AHBERR1:
+          name: AHBERR
+        SETUP1:
+          name: SETUP
+        OUTTKNEPDIS1:
+          name: OUTTKNEPDIS
+        STSPHSERCVD1:
+          name: STSPHSERCVD
+        BACK2BACKSETUP1:
+          name: BACK2BACKSETUP
+        OUTPKTERR1:
+          name: OUTPKTERR
+        BNAINTR1:
+          name: BNAINTR
+        PKTDRPSTS1:
+          name: PKTDRPSTS
+        BBLEERR1:
+          name: BBLEERR
+        NAKINTRPT1:
+          name: NAKINTRPT
+        NYEPINTRPT1:
+          name: NYEPINTRPT
+        STUPPKTRCVD1:
+          name: STUPPKTRCVD
+    DOEPTSIZ[1-6]:
+      name: DOEPTSIZ
+    DOEPDMA[1-6]:
+      name: DOEPDMA
+      _modify:
+        DMAADDR1:
+          name: DMAADDR
+    DOEPDMAB[1-6]:
+      name: DOEPDMAB
+      _modify:
+        DMABUFFERADDR1:
+          name: DMABUFFERADDR
+    _modify:
+      DOEPCTL:
+        fields:
+          MPS:
+            bitOffset: 0
+            bitWidth: 11
+            access: read-write
+          USBACTEP:
+            bitOffset: 15
+            bitWidth: 1
+            access: read-write
+          NAKSTS:
+            bitOffset: 17
+            bitWidth: 1
+            access: read-only
+          EPTYPE:
+            bitOffset: 18
+            bitWidth: 2
+            access: read-write
+          SNP:
+            bitOffset: 20
+            bitWidth: 1
+            access: read-write
+          STALL:
+            bitOffset: 21
+            bitWidth: 1
+            access: read-write
+          CNAK:
+            bitOffset: 26
+            bitWidth: 1
+            access: write-only
+          SNAK:
+            bitOffset: 27
+            bitWidth: 1
+            access: write-only
+          SETD0PID:
+            bitOffset: 28
+            bitWidth: 1
+            access: write-only
+          SETD1PID:
+            bitOffset: 29
+            bitWidth: 1
+            access: write-only
+          EPDIS:
+            bitOffset: 30
+            bitWidth: 1
+            access: read-write
+          EPENA:
+            bitOffset: 31
+            bitWidth: 1
+            access: read-write
+      DOEPTSIZ:
+        fields:
+          XFERSIZE:
+            bitOffset: 0
+            bitWidth: 19
+            access: read-write
+          PKTCNT:
+            bitOffset: 19
+            bitWidth: 10
+            access: read-write
+          SUPCNT:
+            bitOffset: 29
+            bitWidth: 2
+            access: read-write
+
+_add:
+  FIFO%s:
+    addressOffset: 0x1000
+    dim: 16
```

Comment:
I'm not really sure why there are 16 FIFO ranges, especially given that there are only 4 DIEPTXF registers (which would appear to support only up to 5 TX FIFOs, given that TX FIFO 0 is configured with the GNPTXFSIZ register).

I got the 16 number here from the C headers in the ESP-IDF.

Reply:
Are you referring to OTG_TRANS_COUNT_WIDTH from https://github.com/espressif/esp-idf/blob/master/components/soc/esp32s3/include/soc/usb_dwc_cfg.h or something else?

Reply:
The structure definitions in soc/usb_struct.h are what I'm referring to here:
https://github.com/espressif/esp-idf/blob/9c99a385adaf6dfc49138f83f97069bac63eef2a/components/soc/esp32s3/include/soc/usb_struct.h#L112

Note that the C definitions of these structures also have some of the same field width bugs that I'm fixing here.  I had reported an issue about that a few months ago:
https://github.com/espressif/esp-idf/issues/12710

As an aside, the reply to that issue indicated that an update to the tinyusb code was in progress that would add DMA support.  I would be interested in any docs or example code about how to do DMA transfers to the USB peripheral.  (It seems a bit unfortunate that Espressif doesn't provide much docs for the USB peripheral, citing NDA with Synposys, while STMicro is able to provide much better docs for the same underlying peripheral.)

Reply:
Oh, interesting, I hadn't really looked through the host-side code in ESP-IDF much before.  I had mostly been looking at he device examples, which all use tinyusb and the usb_struct.h headers rather than usb_dwc_struct.h  The host side code does appear to have some examples using DMA.

Reply:
Okay so I think I understand where the 16 comes from, our configuration has `OTG_NUM_HOST_CHAN = 8`, and all IN EPs have RX FIFOs by default. We also have `OTG_EN_DED_TX_FIFO = 1`, meaning for each channel there is a dedicated TX FIFO too, giving the total of 16. I'm not sure, but I don't think all these FIFOs are used in device mode.

> As an aside, the reply to that issue indicated that an update to the tinyusb code was in progress that would add DMA support. I would be interested in any docs or example code about how to do DMA transfers to the USB peripheral. (It seems a bit unfortunate that Espressif doesn't provide much docs for the USB peripheral, citing NDA with Synposys, while STMicro is able to provide much better docs for the same underlying peripheral.)

Yes, this is a bit unfortunate, the programming of the core is mostly the same between the two, but there are a whole bunch of configuration options which are only documented in code at this point. I just got access to the synopsis docs, so feel free to ask me any questions and I'll try and answer them as best I can (the docs are vast, with multiple levels of dependencies on specific features).